File: metal/sync_lat.m
// sync_lat.m — G1+G2: CPU<->GPU sync latency on M4 Max (GPU phase).
// G1: classic dispatch round-trip (commit + waitUntilCompleted, 1-thread kernel)
// G2: persistent-kernel mailbox — GPU kernel spins on shared-memory word; CPU writes a
//     request seq, GPU echoes it; N handshakes inside ONE kernel launch.
// Hybrid single-stream decode needs ~48-96 syncs/token; budget <20us each.
// Build: clang -O2 -fobjc-arc sync_lat.m -framework Metal -framework Foundation -o sync_lat
#import <Metal/Metal.h>
#import <Foundation/Foundation.h>
#include <stdatomic.h>
#include <mach/mach_time.h>

static double now_us(void) {
    static mach_timebase_info_data_t tb;
    if (!tb.denom) mach_timebase_info(&tb);
    return (double)mach_absolute_time() * tb.numer / tb.denom / 1000.0;
}
static int cmpd(const void *a, const void *b) {
    double x = *(const double*)a, y = *(const double*)b;
    return x < y ? -1 : x > y;
}

int main(void) { @autoreleasepool {
    setvbuf(stdout, NULL, _IONBF, 0);
    id<MTLDevice> dev = MTLCreateSystemDefaultDevice();
    id<MTLCommandQueue> q = [dev newCommandQueue];
    printf("device: %s\n", dev.name.UTF8String);

    NSString *src = @"#include <metal_stdlib>\n"
        "using namespace metal;\n"
        "kernel void knop(device uint* x [[buffer(0)]], uint gid [[thread_position_in_grid]]) {\n"
        "    if (gid == 0) x[0] += 1;\n"
        "}\n"
        "kernel void kmailbox(volatile device atomic_uint* m [[buffer(0)]],\n"
        "                     uint gid [[thread_position_in_grid]]) {\n"
        "    if (gid != 0) return;\n"
        "    uint iters = atomic_load_explicit(&m[2], memory_order_relaxed);\n"
        "    uint last = 0;\n"
        "    for (uint i = 0; i < iters; i++) {\n"
        "        uint req;\n"
        "        do { req = atomic_load_explicit(&m[0], memory_order_relaxed); } while (req == last);\n"
        "        last = req;\n"
        "        atomic_store_explicit(&m[1], req, memory_order_relaxed);\n"
        "    }\n"
        "}\n";
    NSError *err = nil;
    id<MTLLibrary> lib = [dev newLibraryWithSource:src options:nil error:&err];
    if (!lib) { printf("compile error: %s\n", err.description.UTF8String); return 1; }
    id<MTLComputePipelineState> nop = [dev newComputePipelineStateWithFunction:[lib newFunctionWithName:@"knop"] error:&err];
    id<MTLComputePipelineState> pmb = [dev newComputePipelineStateWithFunction:[lib newFunctionWithName:@"kmailbox"] error:&err];
    id<MTLBuffer> tiny = [dev newBufferWithLength:4096 options:MTLResourceStorageModeShared];

    // ---------- G1 ----------
    double t[400];
    for (int i = 0; i < 200; i++) {
        double t0 = now_us();
        id<MTLCommandBuffer> cb = [q commandBuffer];
        id<MTLComputeCommandEncoder> e = [cb computeCommandEncoder];
        [e setComputePipelineState:nop];
        [e setBuffer:tiny offset:0 atIndex:0];
        [e dispatchThreads:MTLSizeMake(1,1,1) threadsPerThreadgroup:MTLSizeMake(1,1,1)];
        [e endEncoding];
        [cb commit]; [cb waitUntilCompleted];
        t[i] = now_us() - t0;
    }
    qsort(t, 200, sizeof(double), cmpd);
    printf("G1 dispatch round-trip: median %6.1f us   p10 %6.1f   p90 %6.1f\n", t[100], t[20], t[180]);

    // ---------- G2 ----------
    id<MTLBuffer> mbox = [dev newBufferWithLength:4096 options:MTLResourceStorageModeShared];
    volatile _Atomic uint32_t *mb = (volatile _Atomic uint32_t *)mbox.contents;
    const uint32_t ITERS = 20000;
    atomic_store(&mb[0], 0); atomic_store(&mb[1], 0); atomic_store(&mb[2], ITERS);

    id<MTLCommandBuffer> cb = [q commandBuffer];
    id<MTLComputeCommandEncoder> e = [cb computeCommandEncoder];
    [e setComputePipelineState:pmb];
    [e setBuffer:mbox offset:0 atIndex:0];
    [e dispatchThreads:MTLSizeMake(1,1,1) threadsPerThreadgroup:MTLSizeMake(1,1,1)];
    [e endEncoding];
    [cb commit];               // kernel now resident + spinning
    usleep(3000);

    static double lat[20000];
    for (uint32_t i = 1; i <= ITERS; i++) {
        double t0 = now_us();
        atomic_store_explicit(&mb[0], i, memory_order_relaxed);       // request
        while (atomic_load_explicit(&mb[1], memory_order_relaxed) != i) {} // echo
        lat[i-1] = now_us() - t0;
    }
    [cb waitUntilCompleted];
    if (cb.status == MTLCommandBufferStatusError)
        printf("G2 kernel ERROR: %s\n", cb.error.description.UTF8String);
    qsort(lat, ITERS, sizeof(double), cmpd);
    printf("G2 mailbox round-trip: median %6.2f us   p10 %6.2f   p90 %6.2f   p99 %6.2f  (n=%u)\n",
           lat[ITERS/2], lat[ITERS/10], lat[ITERS*9/10], lat[ITERS*99/100], ITERS);
    printf("budget: 96 syncs/token x median = %5.3f ms/token\n", 96.0 * lat[ITERS/2] / 1000.0);
    return 0;
} }

File: metal/sync_lat.swift
// sync_lat.swift — G1+G2: CPU<->GPU synchronization latency on M4 Max (GPU phase).
// G1: classic dispatch round-trip (commandBuffer commit + waitUntilCompleted, empty kernel)
// G2: persistent-kernel mailbox — GPU kernel spins on a shared-memory atomic; CPU writes
//     a request seq, GPU echoes it back; N handshakes inside ONE kernel launch.
// The hybrid single-stream engine needs ~48-96 syncs/token; budget <20us each.
// Build: swiftc -O sync_lat.swift -o sync_lat   Run: ./sync_lat
import Metal
import Foundation

guard let dev = MTLCreateSystemDefaultDevice() else { fatalError("no Metal") }
let q = dev.makeCommandQueue()!
print("device: \(dev.name)")

// ---------- G1: classic dispatch round-trip ----------
let empty = """
#include <metal_stdlib>
using namespace metal;
kernel void knop(device uint* x [[buffer(0)]], uint gid [[thread_position_in_grid]]) {
    if (gid == 0) x[0] += 1;
}
"""
let lib1 = try dev.makeLibrary(source: empty, options: nil)
let nop = try dev.makeComputePipelineState(function: lib1.makeFunction(name: "knop")!)
let tiny = dev.makeBuffer(length: 64, options: .storageModeShared)!

var times: [Double] = []
for _ in 0..<200 {
    let t0 = DispatchTime.now().uptimeNanoseconds
    let cb = q.makeCommandBuffer()!
    let e = cb.makeComputeCommandEncoder()!
    e.setComputePipelineState(nop)
    e.setBuffer(tiny, offset: 0, index: 0)
    e.dispatchThreads(MTLSize(width: 1, height: 1, depth: 1),
                      threadsPerThreadgroup: MTLSize(width: 1, height: 1, depth: 1))
    e.endEncoding()
    cb.commit(); cb.waitUntilCompleted()
    times.append(Double(DispatchTime.now().uptimeNanoseconds - t0) / 1000.0)
}
times.sort()
print(String(format: "G1 dispatch round-trip: median %6.1f us   p10 %6.1f   p90 %6.1f",
             times[100], times[20], times[180]))

// ---------- G2: persistent-kernel mailbox ----------
// mailbox layout (uint32 slots, storageModeShared):
//   [0] = request seq (CPU writes)   [1] = response seq (GPU writes)   [2] = iters
let mbox = dev.makeBuffer(length: 4096, options: .storageModeShared)!
let mb = mbox.contents().bindMemory(to: UInt32.self, capacity: 1024)

let persistent = """
#include <metal_stdlib>
using namespace metal;
kernel void kmailbox(volatile device atomic_uint* m [[buffer(0)]],
                     uint gid [[thread_position_in_grid]]) {
    if (gid != 0) return;
    uint iters = atomic_load_explicit(&m[2], memory_order_relaxed);
    uint last = 0;
    for (uint i = 0; i < iters; i++) {
        uint req;
        do { req = atomic_load_explicit(&m[0], memory_order_relaxed); } while (req == last);
        last = req;
        atomic_store_explicit(&m[1], req, memory_order_relaxed);
    }
}
"""
let lib2 = try dev.makeLibrary(source: persistent, options: nil)
let pmb = try dev.makeComputePipelineState(function: lib2.makeFunction(name: "kmailbox")!)

let ITERS: UInt32 = 20000
mb[0] = 0; mb[1] = 0; mb[2] = ITERS
let cb = q.makeCommandBuffer()!
let e = cb.makeComputeCommandEncoder()!
e.setComputePipelineState(pmb)
e.setBuffer(mbox, offset: 0, index: 0)
e.dispatchThreads(MTLSize(width: 1, height: 1, depth: 1),
                  threadsPerThreadgroup: MTLSize(width: 1, height: 1, depth: 1))
e.endEncoding()
cb.commit()   // kernel now spinning on the mailbox

usleep(2000)  // let it arrive
var lat: [Double] = []
let volPtr = UnsafeMutablePointer<UInt32>(mutating: mb)
for i in 1...ITERS {
    let t0 = DispatchTime.now().uptimeNanoseconds
    // publish request
    OSAtomicCompareAndSwap32(Int32(bitPattern: volPtr[0]), Int32(bitPattern: i), UnsafeMutablePointer<Int32>(OpaquePointer(volPtr)))
    // spin for echo
    while volPtr[1] != i { /* spin */ }
    lat.append(Double(DispatchTime.now().uptimeNanoseconds - t0) / 1000.0)
}
cb.waitUntilCompleted()
if cb.status == .error { print("G2 kernel ERROR: \(String(describing: cb.error))") }
lat.sort()
let n = lat.count
print(String(format: "G2 mailbox round-trip: median %6.2f us   p10 %6.2f   p90 %6.2f   p99 %6.2f  (n=%d)",
             lat[n/2], lat[n/10], lat[n*9/10], lat[n*99/100], n))
print(String(format: "budget check: 96 syncs/token x median = %5.2f ms/token", 96.0 * lat[n/2] / 1000.0))
